masterbase/anomaly.py

```diff
@@ -0,0 +1,62 @@
+"""Anomaly detection for demo streams."""
+
+from typing import NamedTuple
+
+import numpy as np
+
+S_hat = np.load("S_hat.npy")
+
+
+def longest_zero_run(data: bytes) -> int:
+    """Get the longest zero run of data."""
+    array = np.frombuffer(data, dtype=np.uint8)
+    zero_mask = array == 0
+    eq_mask = np.empty_like(array, dtype=bool)
+    eq_mask[0] = True
+    eq_mask[1:] = array[1:] == array[:-1]
+    run_mask = zero_mask & eq_mask
+    total_runs = np.cumsum(run_mask.astype(int))
+    run_lengths = total_runs - np.maximum.accumulate(np.where(run_mask, 0, total_runs))
+    return np.max(run_lengths) + 1
+
+
+def likelihood(p, q) -> float:
+    """Determine the likelihood."""
+    return np.exp(np.sum(np.log(p + 1e-5) * q))
+
+
+def nz_markov_likelihood(S_hat, coocs) -> float:
+    """Determine the NZ-Markov likelihood."""
+    S_hat, coocs = map(lambda a: a.reshape(-1)[1:], (S_hat, coocs))
+    S_hat, coocs = map(lambda a: a / a.sum(), (S_hat, coocs))
+    return likelihood(S_hat, coocs)
```

Comment:
What is `S_hat` in reference to scope here? We read `S_hat` from the file up top, so I assume we can omit the `S_hat` argument in this function?

Reply:
see below